### ISSUE TITLE
Minimize to notification area/ taskbar tray

### ISSUE THREAD

**Issue by Garry3064** [filips123/PWAsForFirefox#146]
### Problem Description
For PWAs of Quora, Whatsapp, Instagram, Reddit, Twitter, etc, It helps a lot if these are minimized to system tray so that one can still receive notifications from them, and they do not clutter up the taskbar.

### Proposed Solution
Most probably, to add a checkmark for this feature in the progressive web apps menu in Firefox settings. I have tried the GNOME Extension 'Minimize to tray', but it was buggy and only worked on X11.

### Additional Information
I use Linux (Zorin OS 16)

**Comment by filips123** [comment]
I agree that minimizing PWAs to system tray would be nice. The main problem is that (I think) each OS and DE does this a bit differently so it would need some work.

Until this is implement, you will probably have to use GNOME extensions or try to add it yourself manually.

**Comment by tfiers** [comment]
For the record, on Windows, I use rbtray, which works well with PWAsFF (right click minimize button to send to tray)

**Comment by Unkn0wn-MDCLXIV** [comment]
Any news on this? This would be a great enhancement

**Comment by filips123** [comment]
Not really... If you need this feature, try using third-party apps/extensions like Minimize to tray on GNOME or rbtray on Windows.

**Comment by K4R7IK** [comment]
Hey, is it possible to minimize the apps to tray now. I made webapps for outlook and Microsoft Teams. To receive the mail notifications and teams notifications, I need to keep the apps running.

If there's a workaround to this problem then, please let me know.

**Comment by zakius** [comment]
Thunderbird has this, also there's extension that closes to tray too, porting that code would be much easier than doing all the research again I believe

**Comment by nekename** [comment]
I would love this. I think cross-platform tray icons are mostly a solved problem - there's the `tray-icon` Rust crate maintained by Tauri, for example. (mentioning because I see Rust in this repo)

**Comment by sleeptightAnsiC** [comment]
I did some research on this, so I thought I would dump all my thoughts here. I was tempted to open a new Issue for "_general Tray support_" but I felt like everyone coming here already expects this. Note that I'm not a Webdev guy, I'm just a user who would really like to see this implemented. Here we go:
- AFAIK, system Tray support is an unsolved problem in PWA implementations. No browser supports this in ~~reliable~~ **ANY** way, so this project would have to emulate it somehow.
- There was a W3C meeting that discussed this. For all links, see: https://github.com/w3c/tpac2024-breakouts/issues/41
- Redditor found a way to emulate something similar on Chromium, but this seems to be very hacky and won't work in some cases, see: https://www.reddit.com/r/PWA/comments/1509s9k/psa_using_pwas_on_a_linux_desktop/
- There is more to "Tray icon" than just indicating an app being minimized:
  - the app should still be alive when minimized and produce/receive updates
  - the icon should dynamically change (e.g. when getting notification, just like the "urgent" tab does)
  - LMB click on the icon should activate/raise the app (obviously?)
  - RMB click on the icon should create a context menu specific to the app (optional but most apps do this)
  - I would expect that an app taking a spot on Tray, should have only one running instance (aka you can't open multiple windows for it)
  - I would expect that an app doing something unusual (like recording the audio or screencasting) would have some indicator about it next to the icon
- There is a "Badging API" that could be reused for deciding what should appear on Tray icon:
  - (I guess, PWA implementations already do this for "taskbar" icon?)
  - https://developer.mozilla.org/en-US/docs/Web/Progressive_web_apps/How_to/Display_badge_on_app_icon
  - https://developer.mozilla.org/en-US/docs/Web/API/Badging_API
- There is a "shortcuts" member in the webapp manifest, that can be used for generating Tray context menu:
  - https://developer.mozilla.org/en-US/docs/Web/Progressive_web_apps/How_to/Expose_common_actions_as_shortcuts
  - https://developer.mozilla.org/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/shortcuts
  - I've tested how shortcuts work when clicked from the taskbar (Youtube PWA under Chromium on Windows11) and it seems to be opening entirely new window, which may make sense for "pinned" desktop entry, but would be totally unexpected when used from Tray menu
- Possibly related to:
  - https://github.com/filips123/PWAsForFirefox/issues/678
  - https://github.com/filips123/PWAsForFirefox/issues/688
  - https://github.com/filips123/PWAsForFirefox/issues/651
  - https://github.com/filips123/PWAsForFirefox/discussions/373
  - https://github.com/filips123/PWAsForFirefox/discussions/320

Hope you don't mind my tiny little spam...

**Comment by filips123** [comment]
Yeah, I agree that it would be useful. The main problem (apart from that I don't have much time currently) is that it's quite complex feature, because as you've said it should integrate with autostart, notifications, etc. Initially, I planned to do all that as part of a big rewrite of the UserChrome code, but that is also a lot of work, so maybe I'll try to implement at least some features first.
